keras_cv/metrics/coco/mean_average_precision.py

```diff
@@ -0,0 +1,316 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+from keras_cv.metrics.coco import utils
+from keras_cv.utils import bbox
+from keras_cv.utils import iou as iou_lib
+
+
+class COCOMeanAveragePrecision(tf.keras.metrics.Metric):
+    """COCOMeanAveragePrecision computes an approximation of MaP.
+
+    Args:
+        class_ids: The class IDs to evaluate the metric for.  To evaluate for
+            all classes in over a set of sequentially labelled classes, pass
+            `range(num_classes)`.
+        iou_thresholds: IoU thresholds over which to evaluate the recall.  Must
+            be a tuple of floats, defaults to [0.5:0.05:0.95].
+        area_range: area range to constrict the considered bounding boxes in
+            metric computation. Defaults to `None`, which makes the metric
+            count all bounding boxes.  Must be a tuple of floats.  The first
+            number in the tuple represents a lower bound for areas, while the
+            second value represents an upper bound.  For example, when
+            `(0, 32**2)` is passed to the metric, recall is only evaluated for
+            objects with areas less than `32*32`.  If `(32**2, 1000000**2)` is
+            passed the metric will only be evaluated for boxes with areas larger
+            than `32**2`, and smaller than `1000000**2`.
+        max_detections: number of maximum detections a model is allowed to make.
+            Must be an integer, defaults to `100`.
+        recall_thresholds: The list of thresholds to average over in the MaP
+            computation.  List of floats.  Defaults to [0:.01:1].
+        num_buckets: num_buckets is used to select the number of confidence
+            buckets predictions are placed into.  Instead of computation MaP
+            over each incrementally selected set of bounding boxes, we instead
+            place them into buckets.  This makes distributed computation easier.
+            Increasing buckets improves accuracy of the metric, while decreasing
+            buckets improves performance.  This is a tradeoff you must weight
+            for your use case.  Defaults to 10,000 which is sufficiently large
+            for most use cases.
+
+    Usage:
+
+    COCOMeanAveragePrecision accepts two Tensors as input to it's `update_state` method.
```

Comment:
Nit: method names should have `()` at the end when referring to them (to mark them as functions/methods)